### ISSUE TITLE
Support for xUnit 2.3

### ISSUE THREAD

**Issue by brutaldev** [axodox/AxoTools#129]
Please add support for xUnit 2.3

**Comment by 304NotModified** [comment]
What's now not working with xUnit 2.3? What needs to be done?

**Comment by 304NotModified** [comment]
I guess strong naming issues

**Comment by jotatoledo** [comment]
any ETA for this feature? @axodox 
BTW, thanks for this great tool!

**Comment by axodox** [comment]
AxoCover was originally made for MSTEST for my personal use. Xunit was requested by community, but it to work properly, I needed to fork its runner and add a small change. This same change was requested by others before me  multiple times in fact and was always rejected. So I need to keep the fork alive.

The other issue is that Xunit runner and library is only working if they are the same version, unlike nunit and mstest, which works even with different combinations quite well.

Updating the fork to Xunit 2.3 is not much work for me, if I break the 2.2 compatibility. To keep both 2.2 and 2.3 however, I need to have 2 forks for the 2 versions and possibly even more later. And for this I need to design a solution which can support this versioning. Or look for other possible solutions.

I don't have much free time either. So if you are willing to help I can point you what needs to be done.

**Comment by 304NotModified** [comment]
> The other issue is that Xunit runner and library is only working if they are the same version, unlike nunit and mstest, which works even with different combinations quite well.

Is this because of strong naming? 

> I don't have much free time either. So if you are willing to help I can point you what needs to be done.

Where is the fork located?

**Comment by 304NotModified** [comment]
> Updating the fork to Xunit 2.3 is not much work for me, if I break the 2.2 compatibility. To keep both 2.2 and 2.3 however, I need to have 2 forks for the 2 versions and possibly even more later. And for this I need to design a solution which can support this versioning. Or look for other possible solutions.

I think it's OK to support only the latest version for now. Especially if we could install older versions of AxoCover

**Comment by axodox** [comment]
No, it is because the internal changes between minor versions are substantial, e.g. dll name changes, namespace, class renames, interface incompatibility etc. AxoCover uses the runner for visual studio internally, the break is between the different versions of the runner vs. the lib. this also causes the tests sometime to not show up in VS either, e.g. when you don't update the runner etc.

Fork is at: https://github.com/axodox/visualstudio.xunit

**Comment by axodox** [comment]
The changes are on the AxoCover branch, you will need to pull the new code from the actual xunit remote (so need to add a remote in git), then you merge the changes, build. After that a nugget package is created and AxoCover is updated to it. Then it will support 2.3, but not 2.2 anymore.

**Comment by Siderite** [comment]
xUnit 2.2.0 breaks some of the tests on my project. 2.3.1 works great. Can you tell me what I need to change in AxoCover to support 2.3.1 on my machine? Thanks!

**Comment by axodox** [comment]
You need to update the fork: https://github.com/axodox/visualstudio.xunit to xUnit 2.3.1 then build AxoCover with it.

**Comment by Panzki** [comment]
Did anyone had success using axocover with xUnit 2.3.?

@axodox Could you please elaborate? I'm not entirely sure how to proceed.

**Comment by adamralph** [comment]
@axodox it may be better to close this issue as a "won't fix", with a clear statement on your position which as I understand it, is:

> AxoCover supports xunit only up to version 2.2. There are no plans to support later versions of xunit. If you would like to add support for later versions of xunit, please fork the project.

**Comment by axodox** [comment]
@adamralph It is not there are no plans or similar, but I have put the project on the parking lane for a while now, as it consumed too much of my free time for years. Another reason is that most of the development I do right now at work either web based or in C++, and the .Net part uses MSTest V1. I understand your frustration. If someone updates AxoCover support, I will integrate it.

**Comment by axodox** [comment]
@Panzki You should compare the AxoCover and master branches of the mentioned VS xunit plug-in fork. Those changes need to be migrated to the version of XUnit you need. Then we can package it with AxoCover for the updated support. It should be doable in less than a day.

**Comment by axodox** [comment]
@Panzki https://github.com/axodox/visualstudio.xunit is used to generate a nuget package, which is used inside AxoCover. So if you have migrated to the latest version, simply generate the nuget package and put into AxoCover dependencies (or for testing just copy the DLLs). Theoretically this should suffice, if there is a difference in the file list, you might need to check the code in AxoCover providing Xunit support: https://github.com/axodox/AxoCover/tree/master/AxoCover/Models/Testing/Adapters

**Comment by adamralph** [comment]
> ...I have put the project on the parking lane for a while now, as it consumed too much of my free time for years. Another reason is that most of the development I do right now at work either web based or in C++...

Perhaps it's time to hand over the reigns to someone else?

**Comment by axodox** [comment]
I am open to that too, if there is someone willing to take over the work.

**Comment by adamralph** [comment]
@axodox I took the liberty of raising a PR over at [maintainers-wanted](https://github.com/pickhardt/maintainers-wanted/pull/35).

**Comment by raytangRT** [comment]
@adamralph sorry, I did not read the PR properly, I thought you are the one who did a PR on `this` project and make it support 2.3

**Comment by 304NotModified** [comment]
@VinS101 Please do 👍

**Comment by jsabrooke** [comment]
@VinS101 any chance you got anywhere with this?

**Comment by VinS101** [comment]
@jsabrooke Haven't started any work yet. It's in my backlog.

**Comment by VinS101** [comment]
I'm not sure what happened, but axoCovered started working for me. I have xUnit 2.4.

**Comment by VinS101** [comment]
@malcolmarobertson  The only thing I can think of: I updated my VS2017 to version 15.9.6 a couple weeks ago.

These are my axoCover settings: 
![image](https://user-images.githubusercontent.com/8950747/53197884-27a09c00-35e9-11e9-826c-c2d18bf5a2ec.png)

**Comment by biancarus88** [comment]
hey @VinS101 . Unfortunately just tried upgrading VS2017 to 15.9.7, but still the tests are not recognized. I am using xUnit 2.4.1 .. 
Did anyone else manage to get this working?

**Comment by malcolmarobertson** [comment]
I got it working using the settings below, but first I followed steps to register OpenCover dll. 
1. Install OpenCover on your Test project from Package Manager
2. Open CMD as admin
3. navigate to new packages\OpenCover directory
4. run \packages\OpenCover.4.7.922\tools\x64>regsvr32 /i OpenCover.Profiler.dll (use x86 if you are using 32bit)

Source for OpenCover register: 
https://github.com/axodox/AxoCover/issues/172

Then my settings look like this:
![image](https://user-images.githubusercontent.com/17067089/53255181-6b68d380-36c5-11e9-9893-5dcf487c8d95.png)
![image](https://user-images.githubusercontent.com/17067089/53255697-6d7f6200-36c6-11e9-8b1c-3dae2baf5e0b.png)

**Comment by biancarus88** [comment]
hey @malcolmarobertson . sorry for the late answer. I have registered the OpenCover.dll as you suggested and worked fine, changed the settings to match yours as well, tried to force it both x64 and x86, but no luck for me. I am using VS2017 Professional, just updated to 15.9.8. You are using VS2017 15.9.6?

**Comment by malcolmarobertson** [comment]
I don't think VS version is an issue, when you build your project and then
reopen AxoCover are there any errors in the output pane? You may need to
select AxoCover from the output window dropdown.

**Comment by nudemeth** [comment]
I've updated xUnit dependencies and compiled the project to support version 2.4.1. Tested in my local VS 2017 and have no problems so far. Uploaded [link](https://drive.google.com/file/d/12iDN9Bg2xWgRG1knui3IY4gsIvZTrXfh/view?usp=sharing).

**Comment by NickGowdy** [comment]
Hey @axodox , I had problems getting AxoCover to detect my tests with xUnit 2.4 and only got it to work by downgrading to xUnit 2.2. Is there any future plans for xUnit support for the latest version (2.4)?
